### ISSUE TITLE
Timeout on /wallet/payment/send 

### ISSUE THREAD

**Issue by kushti** [ergoplatform/ergo#1885]
/wallet/payment/send (url/wallet/payment/send method:POST params:[map[address:9... value:+10000000000]])

returns often: `The server was not able to produce a timely response to your request`

The wallet has default settings, ~700 utxos, ~27K addresses

**Comment by pragmaxim** [comment]
Yeah this can happen if readers are not ready (`readersHolder ? GetReaders`)... At least I find it high likely as the timeout is set to 1 minute. Maybe we should `Http().newServerAt` when the readers are ready? As most of the endpoints depends on at least on state component.

**Comment by pragmaxim** [comment]
I cannot reproduce it :cry: this looks suspicious to me : 
https://github.com/ergoplatform/ergo/blob/8066c8427d7e079b1736d072c190613c897696d6/ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ReplaceCompactCollectBoxSelector.scala#L53-L57

We grab by default `maxInputs * BoxSelector.ScanDepthFactor = 30 000`  boxes and all the recursive `replace` method that does some very intense ops like `partition` on possibly huge collections :shrug:

**Comment by Ergologica** [comment]
Before anyone spends time on this: **#1887 says `Closes #1885`, and it was merged by @kushti on 2022-10-27** — the same day @pragmaxim posted the comment above. This issue stayed open anyway, and still carries `B-500 SigUSD`.

I went looking for the timeout on current master and measured the component everyone points at. `ReplaceCompactCollectBoxSelector` in isolation — no LevelDB, no node, boxes built in memory, `maxInputs = 100`, `optimalInputs = 3`, tail capped at `maxInputs * ScanDepthFactor` = 30 000:

| boxes | selection within `maxInputs` (`collectDust`) | selection over `maxInputs` (`replace`) |
|---|---|---|
| 1 000 | 20 ms | 43 ms |
| 10 000 | 6 ms | 761 ms |
| 30 000 | **12 ms** | 1 916 ms |

Dust collection is flat — 12 ms over 30 000 boxes, against the 20 seconds #1887 reported. Those four fixes (`span` instead of `partition`, `HashSet` instead of `List.contains`, hoisting `valueOf`, no appending to a huge list) are in the code today and they work. `replace` is still slowish at ~2 s, but it cannot grow beyond that: the tail is truncated at 30 000 by construction, whatever the wallet size.

So **the reported symptom may simply be fixed and the issue left open by accident.** If so this should be closed and the bounty retired, and I would rather say that than quietly claim it.

---

If the timeout is still seen in the wild, then it is not the selector, and the arithmetic points somewhere specific.

`application.conf` line 386, under `akka.http.server`, is `request-timeout = 1 minute` — and *"The server was not able to produce a timely response to your request"* is akka-http's default response for it. Two seconds of selector does not reach a minute. The question is what stops the response entirely.

Every wallet call goes through `withWalletOp` (`WalletApiOperations.scala:71`):

```scala
onSuccess((readersHolder ? GetReaders).mapTo[Readers].flatMap(r => op(r.w)))(toRoute)
```

`ErgoReadersHolder` first, then `walletActor ? GenerateTransaction`. Both run on `api-dispatcher` (`ErgoWalletActor.scala:513`, `ErgoReadersHolder.scala:72`), which is:

```
fork-join-executor { parallelism-min = 1, parallelism-max = 2 }
throughput = 4
```

Two threads. And `ErgoWalletActor` is not an API actor in any real sense — one `receive` handles `ScanOnChain` per block, `ScanOffChain` per mempool transaction, `ScanInThePast` (which re-sends itself per block while catching up), `Rollback`, *and* `GenerateTransaction`, holding its thread for up to four of them at a time. With ~27K addresses that scanning is real work, on a pool of two, shared with the one actor every wallet API call must reach first.

Which is @pragmaxim's *first* comment, not his second — "readers are not ready" being `ErgoReadersHolder` waiting for a thread rather than the node still starting. The docstring on `GlobalConstants.ApiDispatcher` even says it exists "to avoid clashing between blockchain processing and API actors"; the actor doing the heaviest blockchain-derived work is on it.

I have not reproduced the reporter's node, so that last part is a mechanism, not a proven cause.

**So, two questions rather than a PR:** should this be closed as fixed by #1887? And if not, is moving `ErgoWalletActor` onto its own dispatcher the direction you would want? I will write it with a test that pins the starvation, but I would rather ask first than hand you a config opinion.

**Comment by Ergologica** [comment]
Asking before anyone picks this up as a bounty: is there anything left here?

@pragmaxim's #1887, "i1885 wallet timeout repro", says "Closes #1885" and was merged into `v5.0.2` in October 2022. The four things it changed are all on master today, in `ReplaceCompactCollectBoxSelector`:

- `val boxesToDrop = mutable.HashSet.empty[T]` (line 140) instead of a list `contains`
- `val (dropped, remain) = toDrop.span { ... }` (line 153) instead of `partition`
- `candValue` hoisted out of the loop (line 152) instead of calling `BoxSelector.valueOf` repeatedly

So the dust collection that took 20 seconds is not the code that runs now. The issue is still open and still carries a 500 SigUSD bounty in the registry, which is why I am asking rather than assuming: is the timeout still reproducible with 700 UTXOs and 27k addresses on a current node, or did this get fixed and simply never closed?
